<h1><b> Predicting whether or not someone will have diabetes </b></h1>

In [43]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split #for splitting the data into testing and training
from sklearn.preprocessing import StandardScaler # this is used to make sure there's no basis for large numbers
from sklearn.neighbors import KNeighborsClassifier #the tool for classifing
from sklearn.metrics import confusion_matrix #for testing
from sklearn.metrics import f1_score #for testing
from sklearn.metrics import accuracy_score #for testing

In [44]:
dataset = pd.read_csv(r'C:\Users\HP\Desktop\Desktop\python\ML algorithms\KNN_Dataset.csv')
dataset.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [45]:
#Removing the 0s since it would be an error for most columns

In [46]:
# we're replacing the zeros with numPy NULL and then creating the mean and replacing the NULLS with the mean
# if we don't have the data aka NULL, we're gonna subsitute the missing data with the average

zeros= ['Glucose', 'BloodPressure', 'SkinThickness', 'BMI', 'Insulin']

for column in zeros:
    dataset[column] = dataset[column].replace(0, np.NaN)
    mean= int(dataset[column].mean(skipna=True))
    dataset[column] = dataset[column].replace(np.NaN, mean)

In [47]:
# Now that we've dealt with any potential nulls, we can start the training:

In [48]:
#split the data:
X= dataset.iloc[:, 0:8] # 0:8 from column Pregnancies-Age, exculding the outcome
y= dataset.iloc[:, 8] # Selecting only the last column, the outcome 
X_train, X_test, y_train, y_test = train_test_split(X,y, random_state=0, test_size=0.2)
#the test size means that we're using 20% of the data to test with it later

In [20]:
# Now we;re gonna feature scale using StandardScaler
# The reason is that we want to standarize our data from -1 to 1 instead of a lot numbers 
# and we're only gonna use it on the training data

In [19]:
sc_X = StandardScaler()
X_train= sc_X.fit_transform(X_train)
X_test= sc_X.transform(X_test)

In [21]:
# Fit the data into the ML model

In [26]:
import math
math.sqrt(len(y_test)) # Since we're gonna use KNN to decide how to classify, we don't want EVEN numbers bc that would be like a tie in voting

12.409673645990857

In [49]:
# we subtracted 1 from the above number (12-1=11) and used that as our n_neighbors, bc odd numbers are the way to go when comes to "voting"
#p=2 bc we want to check if they're diabetic or not, so 2 options
#euclidean is the most common meteric
classifer = KNeighborsClassifier(n_neighbors=11, p=2, metric='euclidean') 

In [34]:
classifer.fit(X_train, y_train)

KNeighborsClassifier(metric='euclidean', n_neighbors=11)

In [35]:
y_pred = classifer.predict(X_test)
y_pred

array([1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
      dtype=int64)

In [36]:
cm= confusion_matrix(y_test, y_pred)
print(cm)

[[92 15]
 [20 27]]


In [50]:
# 92 true positives, 15 false postives 
# 20 false negative, 27 true negatives

In [42]:
print(f1_score(y_test, y_pred))

0.6067415730337079


In [40]:
print(accuracy_score(y_test, y_pred))

0.7727272727272727


<h2> We were able to predict with 72% accuracy whether or not someone is diabetic using KNN algorithm </h2>